### ISSUE TITLE
Cache the structure of the tables so they can be displayed faster

### ISSUE THREAD

**Issue by abodnar** [Sequel-Ace/Sequel-Ace#1571]
Each time you go to view the structure of the table, Sequel Ace has to reach out the server to get what the current structure is and the UI at the time displays a loading screen till it gets a response. In most cases, the structure hasn't changed, so the loading UI is really unnecessary.

Ideally, when you connect to a database, a thread could be used to retrieve the structure for each table and cache the results of that. That way when a user goes to the structure view, the cache of that table can be retrieved to quickly give the user the view of the table. In the background, a thread is spawned that goes to retrieve the structure to see if it has changed since the cache was created. If no changes, then nothing needs to happen. If it does, then a message or something is displayed to let the user know why the view has been updated with the new structure. Also any time a user makes an alter in Sequel Ace, we would want to trash the cached structure and update it after the change occurs.

**Comment by Jason-Morcos** [comment]
> In most cases, the structure hasn't changed

I think this is the tough part - it's tough to know when the structure _has_ changed.


> Each time you go to view the structure of the table, Sequel Ace has to reach out the server to get what the current structure is and the UI at the time displays a loading screen till it gets a response. In most cases, the structure hasn't changed, so the loading UI is really unnecessary.
> 
> Ideally, when you connect to a database, a thread could be used to retrieve the structure for each table and cache the results of that. That way when a user goes to the structure view, the cache of that table can be retrieved to quickly give the user the view of the table. In the background, a thread is spawned that goes to retrieve the structure to see if it has changed since the cache was created. If no changes, then nothing needs to happen. If it does, then a message or something is displayed to let the user know why the view has been updated with the new structure. Also any time a user makes an alter in Sequel Ace, we would want to trash the cached structure and update it after the change occurs.

Two big concerns here:
1. This feels really heavy in large databases where you might only connect to a DB and go to one or two tables. I don't want to get the structure of 100 tables if you only might look at 1-2.
2. Security auditing - this makes it look like the user is "looking" at every table when in reality they might not be. This could open up our users to liability if a data breach occurred at their company.

> In the background, a thread is spawned that goes to retrieve the structure to see if it has changed since the cache was created. 

I feel like this opens us up to crashes if the structure _did_ change and the user tries to edit the structure or something based on an old cache?


> Also any time a user makes an alter in Sequel Ace, we would want to trash the cached structure and update it after the change occurs.

For sure, but what about if changes were made by another client/the server? In my workflows, that's certainly the way more common case.


In summary, I think I'm opposed to this change - I think it opens us up to issues with data being out of sync and the only benefit is a slightly faster load of structure when table structure is hit. The benefits don't seem worth the risks to me here.

**Comment by Jason-Morcos** [comment]
If others have thoughts here, please weight in. My vote is to close this one

**Comment by stychos** [comment]
Only as an option, maybe. As @Jason-Morcos wrote, my case is huge databases with thousands of tables, it will mess things and eat resources.

**Comment by samuelbradshaw** [comment]
I agree with the concerns of pre-caching all tables for large databases, and trying to edit database structure without the latest. However, you could probably avoid most of the issues by:
- Only caching the tables that are visited while connected to the database
- Clearing the cache when disconnecting from the database
- Allowing viewing but not editing the table structure, until the background thread comes back with the latest

If you went down this path, you would probably also want to debounce requests for the latest table structure, if response time is consistently slow, to avoid redundant background threads if a user leaves and comes back multiple times.

**Comment by abodnar** [comment]
> 1. This feels really heavy in large databases where you might only connect to a DB and go to one or two tables. I don't want to get the structure of 100 tables if you only might look at 1-2.

I think the best way to make sure it's not a resource hog is to have a way to enable it either programmatically or at the connection level.

> 2. Security auditing - this makes it look like the user is "looking" at every table when in reality they might not be. This could open up our users to liability if a data breach occurred at their company.

I mean, if you already have access to the tables...but honestly if that is happening, they are going to be looking for a pattern of more than just doing `desc <table>`. You could just query the tables in the INFORMATION_SCHEMA database for the information so you don't have to go to every single table. 

> I feel like this opens us up to crashes if the structure _did_ change and the user tries to edit the structure or something based on an old cache?
> For sure, but what about if changes were made by another client/the server? In my workflows, that's certainly the way more common case.

I believe Sequel Ace already has this risk, right now. If I have a table structure open and something changes that in the time between when I opened the structure and when I go to make my change. My alter may overwrite whatever change was previously done. Only thing that might stop it is if my change somehow causes an error. But like @samuelbradshaw noted, if you don't allow an edit till the background verifies the current structure, then that should stop a change from altering an outdated structure. 

I think @samuelbradshaw noted some good rules to put around this idea.
